### PR TITLE
enhancement(statsd source): comply with component spec

### LINKED ISSUE #10077: Instrument the `statsd` source with event processing metrics
Per #9326

### PR DESCRIPTION
**Pull request by jdrouet**
- rename `StatsdEventReceived` and update tags
- rename `StatsdInvalidRecordError` and update log, events and tags
- add `BytesReceived`
- add `StreamClosedError`
- update `out.send` to `out.send_all`

closes #10077

Signed-off-by: Jérémie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

File: src/internal_events/statsd_source.rs
Comment (001wwang) [inline src/internal_events/statsd_source.rs:15]:
```suggestion
            message = "Events received.",
```
(for consistency)

File: src/sources/statsd/mod.rs
Comment (001wwang) [inline src/sources/statsd/mod.rs:155]: It seems like this Deserializer is also used for the `TCP` and `Unix` modes ([example](https://cs.github.com/vectordotdev/vector/blob/d19f8bc2087ae324b7636f19ee47a2299e49659f/src/sources/statsd/mod.rs#L228)), so the protocol tag doesn't always fit here. Additionally, the underlying TCP source utility logic already emits `TcpBytesReceived` ([see here](https://cs.github.com/vectordotdev/vector/blob/4629f4ad8c5e9d124f29249835b3c7a0592f4f25/src/sources/util/tcp/mod.rs#L266)), so it looks like this would lead to accidental double emission.

Comment (jdrouet) [reply src/sources/statsd/mod.rs]: 👍 I remove it

File: src/sources/statsd/mod.rs
Comment (jszwedko) [inline src/sources/statsd/mod.rs:158]: Similar to other PR comments, I think the `byte_size` being used here is wrong. It should be using `ByteSizeOf::size_of` to emit a counter of in-memory bytes.

**Review by jszwedko — CHANGES_REQUESTED** [review]
Another comment about `component_received_event_bytes_total` 😄

File: src/sources/statsd/mod.rs
Comment (001wwang) [reply src/sources/statsd/mod.rs]: Sorry for the late reply here. I think removing here is correct to fix the double emission for `TCP` mode, but does this leave the `UDP` and `Unix` modes without a `BytesReceived` event? There might be some overlap here with what you decide to do in #11124 (think some underlying logic is shared) to instrument `unix_*` modes.

**Comment by jdrouet** [comment]
In [this commit](https://github.com/vectordotdev/vector/pull/11124/commits/c3bc11d2860ddaf65439c7e7e21b528e3b8a0845), I move the emission of the `EventsReceived` upstream, which impacts `statd` source code, so better merge #11124 first.

File: src/sources/statsd/mod.rs
Comment (jdrouet) [reply src/sources/statsd/mod.rs]: the `UDP` mode has the `BytesReceived` now, and the `Unix` mode as well, only with `datagram`. But shit should probably be discussed in #11124.

File: src/sources/statsd/mod.rs
Comment (001wwang) [inline src/sources/statsd/mod.rs:158]: With #11124 moving `EventsReceived` into places like `build_unix_data_stream` and `TcpSource`, I believe this will cause double counting issues. 

However, if this is removed, statsd `UDP` mode looks like it'll be missing `EventsReceived`; it may also still be missing `BytesReceived`? Sorry if I'm missing it somewhere. It would be helpful for reviewing these instrumentation PRs if you could point out where relevant internal events are emitted since it's not always clear from the changes.

**Comment by jszwedko** [comment]
I'll take another look at this one once https://github.com/vectordotdev/vector/pull/11124 is in.

File: src/sources/statsd/mod.rs
Comment (jdrouet) [reply src/sources/statsd/mod.rs:158]: I guess this PR need to be rebased on top of what is done in #11124. I'll wait that is get merged to proceed 😉

File: src/internal_events/statsd_source.rs
Comment (jszwedko) [inline src/internal_events/statsd_source.rs:27]: Per #11317 I think we should turn this into an error code.

**Review by jszwedko — COMMENTED** [review]
Just one question about the `error` tag, otherwise looks good!

File: src/internal_events/statsd_source.rs
Comment (bruceg) [inline src/internal_events/statsd_source.rs:18]: Will this be updated to a constant once #11334 lands?

**Review by bruceg — APPROVED** [review]
LGTM pending one outstanding tweak

**Pull request merged**
